File: Python/BigNumber.py
from util.CheckTime import TimeCheckUtil

def main():

    # Test Vector
    # line 1: 5 8 3
    # line 2: 2 4 5 4 6

    print('### BigNumber.py ###') # Chap3. Exercise 3-2.
    tc = TimeCheckUtil() # init start time

    line1 = '5 8 3'
    line2 = '2 4 5 4 6'

    #### Code ####

    try :
        result = find_result(line1, line2)
    except ValueError as e:
        print('잘못된 값이 입력되었습니다.', e)
        return 1

    print('Result : ', str(result))

    #### Code ####

    tc.print_exec_time() # print time duration

    return 0

def find_result(info, data):
    info = list(map(int, info.split())) # 입력 처리
    data = list(map(int, data.split())) # 입력 처리

    N = info[0] # data list 크기. (2 <= N <= 1,000)
    M = info[1] # 덧셈 연산 횟수. (1 <= M <= 10,000)
    K = info[2] # 최대 연속 덧셈 가능 횟수. 단, 값이 같아도 다른 인덱스에 있으면 다른 수로 간주. (1 <= K <= 10,000)

    # 입력값 검증
    if 2 > N or N > 1000:
        raise ValueError
    
    if 1 > M or 10000 < M:
        raise ValueError
    
    if 1 > K or 10000 < K:
        raise ValueError
    
    data.sort(reverse=True)

    biggest_number = data[0]
    second_biggest_number = data[1]

    k = 0
    sum = 0

    for m in range(M):
        if k < K:
            sum += biggest_number
            k += 1
        else :
            sum += second_biggest_number
            k = 0
    
    return sum

main()

File: C/BigNumber.c
#include <stdio.h>
#include <string.h>

int main(void) {
    char info[5] = "5 8 3";
    char *input = "2 4 5 4 6";

    /* 입력값 검증 */
    
    int N = info[0] - '0';
    if ( 2 > N || 1000 < N ) {
        return 1;
    } 

    int M = info[2] - '0';
    if ( 1 > M || 10000 < M ) {
        return 1;
    }

    int K = info[4] - '0';
    if ( 1 > K || 10000 < K ) {
        return 1;
    }
    
    int inputLen = strlen(input);

    int inputArr[N];
    int number;

    inputArr[0] = input[0] - '0';
    int inx = 1;

    for (int i = 1; i < inputLen; i++) {
        if (i%2 == 0) {
            number = input[i] - '0';

            // 내림차순 정렬
            for(int j = 0; j < inx; j++) {
                if(number > inputArr[j]) {
                    int tmp = inputArr[j];
                    inputArr[j] = number;
                    number = tmp;
                } 
            }
            inputArr[inx] = number;
            inx++;
        }
    }

    int biggestNumber = inputArr[0];
    int secondBiggestNumber = inputArr[1];

    int count = 0;
    int sum = 0;
    for (int i = 0; i < M; i++) {
        if (count < K) {
            sum += biggestNumber;
            count++;
        }
        else {
            sum += secondBiggestNumber;
            count = 0;
        }
    }

    printf("Result : %d\n", sum);

    return 0;
}